core/src/main/scala/com/raphtory/components/querytracker/QueryProgressTracker.scala

```diff
@@ -65,89 +60,92 @@ class QueryProgressTracker(
 
-  def isJobDone(): Boolean =
-    jobDone
+        logger.info(s"Job $jobID: Running query, processed $perspectivesProcessed perspectives.")
 
-  def waitForJob() =
-    //TODO as executor sleep
-    while (!jobDone)
-      Thread.sleep(1000)
+      case JobDone                  =>
+        logger.info(
+                s"Job $jobID: Query completed with $perspectivesProcessed perspectives " +
+                  s"and finished in ${System.currentTimeMillis() - startTime} ms."
+        )
+
+        jobDone = true
+    }
+
+  override def run(): Unit = {
+    logger.info(s"Job $jobID: Starting query progress tracker.")
 
-  // Starts the query tracker pulsar consumer
-  override def run(): Unit =
     cancelableConsumer = Some(
             pulsarController.startQueryTrackerConsumer(jobID, messageListener())
     )
+  }
 
-  // Stops the query tracker consumer
-  def stop(): Unit =
+  override def stop(): Unit =
     cancelableConsumer match {
       case Some(value) =>
         value.close()
       case None        =>
     }
 
-  // Handles message to process the {s}`Perspective` received in case the query is in progress, or {s}`JobDone` if the query is complete
-  override def handleMessage(msg: QueryManagement): Unit =
-    msg match {
-
-      case p: Perspective =>
-        val perspectiveDuration = System.currentTimeMillis() - perspectiveTime
+  def getJobId: String =
+    jobID
 
-        if (p.window.nonEmpty)
-          logger.info(
-                  s"Job '$jobID': Perspective '${p.timestamp}' with window '${p.window.get}' finished in $perspectiveDuration ms."
-          )
-        else
-          logger.info(
-                  s"Job '$jobID': Perspective '${p.timestamp}' finished in $perspectiveDuration ms."
-          )
+  def getLatestPerspectiveProcessed: Option[Perspective] =
+    latestPerspective
 
-        perspectiveTime = System.currentTimeMillis
-        perspectivesProcessed += 1
-        latestPerspective = p
-        perspectivesList += p
-        perspectivesDurations += perspectiveDuration
+  def getPerspectivesProcessed: List[Perspective] =
+    perspectivesList.toList
 
-        logger.info(s"Job $jobID: Running query, processed $perspectivesProcessed perspectives.")
+  def getPerspectiveDurations: List[Long] =
+    perspectivesDurations.toList
 
-      case JobDone        =>
-        logger.info(
-                s"Job $jobID: Query completed with $perspectivesProcessed perspectives " +
-                  s"and finished in ${System.currentTimeMillis() - startTime} ms."
-        )
+  def isJobDone: Boolean =
+    jobDone
 
-        jobDone = true
+  def waitForJob(): Unit = {
+    val sleepTimeInMillis = 1000
 
-      // TODO Need to re-enable below
-      // close consumer
-      // stop()
+    //TODO as executor sleep
+    while (!jobDone) {
+      logger.debug(s"Job '$jobID': Job not yet finished. Sleeping for $sleepTimeInMillis ms.")
+      Thread.sleep(sleepTimeInMillis)
     }
+
+    this.stop()
```

Comment:
This is better than before but note that `stop()` is only called if we use `waitForJob()`. Pushed a commit to fix this.